### PR TITLE
Eq 781 mandatory first name

### PR DESCRIPTION
**Pull request by benjefferies**
### What is the context of this PR?
* Configuring first name on household composition question to be mandatory
* Fixing generate pages script
* Adding logic to validate we are submitting to the correct page

### How to review 
Selenium tests should pass

### CODE REVIEW

File: app/data/census_individual.json [hunk 1]
```diff
@@ -34,23 +34,23 @@
                                             "label": "First Name",
                                             "mandatory": false,
```
Comment (jonnyshaw89) [inline app/data/census_individual.json]: Should this one also be mandatory?

File: tests/functional/generate_pages.py [hunk 2]
```diff
@@ -90,10 +131,15 @@ def process_answer(answer, page_spec):
         answer_name = generate_camel_case_from_id(answer['id'])
         page_spec.write(_write_month_year_date_answer(answer_name, answer['id']))
 
-    elif answer['type'] == 'TextField' or answer['type'] == 'Integer' or answer['type'] == 'Textarea':
-        answer_name = generate_camel_case_from_id(answer['id'])
-        page_spec.write(ANSWER_SETTER.replace("{answerName}", answer_name).replace("{answerId}", answer['id']))
-        page_spec.write(ANSWER_GETTER.replace("{answerName}", answer_name).replace("{answerId}", answer['id']))
+    elif answer['type'] == 'TextField' or answer['type'] == 'Integer' or answer['type'] == 'TextArea':
+        if question_type == 'RepeatingAnswer':
+            answer_name = generate_camel_case_from_id(answer['id'])
```
Comment (jonnyshaw89) [inline tests/functional/generate_pages.py]: define this outside of the if so you don't also have to define it in the else

File: tests/functional/spec/census-household.spec.js [hunk 3]
```diff
@@ -83,7 +83,7 @@ describe('Census Household', function () {
 
         // who-lives-here
         PermanentOrFamilyHome.clickPermanentOrFamilyHomeAnswerYes().submit()
-        HouseholdComposition.setPersonName(0, 'John Smith').submit()
+        HouseholdComposition.setFirstName(0, 'John').submit()
```
Comment (jonnyshaw89) [inline tests/functional/spec/census-household.spec.js]: do we have any scenario where we enter middle and last name?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: tests/functional/spec/census-household.spec.js (hunk 3 above)
Comment (benjefferies) [reply tests/functional/spec/census-household.spec.js]: Don't think we need to. We need to test the scenario that the middle/last is optional. Maybe that would be good as an explicit test. As for the second test I am going to delete, it's not really a selenium test, it is unit tested

File: tests/functional/pages/surveys/census/household/address-type.page.js [hunk 4]
```diff
@@ -2,6 +2,11 @@ import MultipleChoiceWithOtherPage from '../../multiple-choice.page'
 
 class AddressTypePage extends MultipleChoiceWithOtherPage {
 
+  isOpen() {
```
Comment (jonnyshaw89) [inline tests/functional/pages/surveys/census/household/address-type.page.js]: Can this not move up the the question page class and then just provide the blockId in each file?

File: tests/functional/spec/census-household.spec.js (hunk 3 above)
Comment (samiwel) [reply tests/functional/spec/census-household.spec.js]: The Jina filter I created should display middle and last names appropriately if only middle and/or last were provided. But the fact that we've added mandatory validation here will help to ensure that we always have something, which is good.

**Comment by samiwel** [comment]
I noticed that the validation messages appears correctly when the Save and continue button is pressed, but when I the add new person button is pressed the error was not displayed yet I was prevented from adding additional household members (presumably due to failed validation).

Does it make sense to ignore the validation check when adding new people? So that users can add as many new people as they like, but fail the validation as soon as they hit save and continue? From a UX point of view, I guess that is more consistent with the way that screen behaves when JS is enabled on that page.

**Comment by benjefferies** [comment]
I think your right Sam, from a UX views they should act the same. I have removed validation on the add/remove action

File: tests/integration/questionnaire/test_questionnaire_household_composition.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        post_data = {
+            "action[add_answer]": ""
+        }
+
+        # When
+        resp = self.client.post(self.household_composition_url, data=post_data)
+
+        # Then
+        self.assertEqual(200, resp.status_code)
+        self.assertNotIn('This field is mandatory.', resp.get_data(True))
```
Comment (benjefferies) [inline tests/integration/questionnaire/test_questionnaire_household_composition.py]: If we returned a status code of 400 for invalid answers we wouldn't need to check the body making testing end points nicer. I know in REST apis status codes are ubiquitous but not sure how it fits in outside REST i.e. will some browsers intervene (probably IE)

**Comment by jonnyshaw89** [comment]
@benjefferies @samiwel i think the removing of validation isn't a good thing. if i click add lots of times, when go previous and say no one lives here, then go forward i get repeated around lots of groups of people with empty names.

File: app/main/views/questionnaire.py [hunk 6]
```diff
@@ -237,18 +237,19 @@ def post_household_composition(eq_id, form_type, collection_id, group_id):
         'group_id': group_id,
         'group_instance': 0,
     }
+    if 'action[add_answer]' in request.form or 'action[remove_answer]' in request.form:
+        questionnaire_manager.build_state(this_block, request.form)
```
Comment (jonnyshaw89) [inline app/main/views/questionnaire.py]: if my request.form contained answers for other blocks would they be applied without being validated. I'm a little worried that we have exposed our answer store to be open to abuse by not validating.

File: tests/functional/generate_pages.py [hunk 7]
```diff
@@ -46,6 +46,13 @@
 
 """
 
+ANSWER_SELECT = r"""  set{answerName}(value) {
```
Comment (jonnyshaw89) [inline tests/functional/generate_pages.py]: ANSWER_SELECT -> ANSWER_SELECTOR?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

**Comment by jonnyshaw89** [comment]
If you add a person with no name, click previous then go forwards again. you can then not ever remove them.

File: tests/functional/pages/surveys/question.page.js [hunk 8]
```diff
@@ -1,10 +1,23 @@
...
+  }
+
   getAlertText() {
     return browser.element('[data-qa="error-body"]').getText()
   }
 
   submit() {
+    if (this.pageName != undefined && !this.isOpen()) {
+    const url = browser.url().value
+        throw new Error(`Expected to be on ${this.pageName} page but was on ${browser.getUrl()}`)
```
Comment (jonnyshaw89) [inline tests/functional/pages/surveys/question.page.js]: Kudos

**Comment by jonnyshaw89** [comment]
I have created an issue (https://github.com/ONSdigital/eq-survey-runner/issues/727) for the scenario i found above. But given that that existed before this PR I'm happy.

**Review by jonnyshaw89 — APPROVED** [review]

**Review by samiwel — APPROVED** [review]
Looks good to me 👍

File: tests/functional/spec/census-household.spec.js [hunk 9]
```diff
@@ -387,6 +387,25 @@ describe('Census Household', function () {
...
+        startCensusQuestionnaire('census_household.json', false, 'GB-WLS')
+
+        // who-lives-here
+        PermanentOrFamilyHome.clickPermanentOrFamilyHomeAnswerYes().submit()
+        HouseholdComposition.submit()
+
+        expect(HouseholdComposition.getErrorMsg()).to.contain('Please enter a name or remove the person to continue')
+    })
+
+    it('Given I enter a first name but no middle or surname, When I save and continue, Then I should see errors that first name is mandatory.', function () {
```
Comment (ajmaddaford) [inline tests/functional/spec/census-household.spec.js]: Should this say no errors?

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: tests/functional/spec/census-household.spec.js (hunk 9 above)
Comment (benjefferies) [reply tests/functional/spec/census-household.spec.js]: Your sharp-eyed, always catching me out on these!

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**